pkg/utils/reconcilename/reconcilename.go

```diff
@@ -0,0 +1,33 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconcilename
+
+import (
+	"context"
+)
+
+type contextKey string
+
+func Inject(ctx context.Context, contextkey string, resourcename string) context.Context {
+	return context.WithValue(ctx, contextKey(contextkey), resourcename)
```

Comment:
Ideally your callers wouldn't need to be aware of the string. The concept of a reconciled object name is general to every controller. There's a bit of a hiccup since some objects can have both names and namespaces, but for us provisioners and nodes are both un-namespaced concepts. I might name it something like
```
injection.WithResourceKey() // returns "name" or "namespace/name"
```
We could also go totally crazy and stuff in a pointer to the full object like
```
injection.WithObject(ctx, provisioner)
injection.GetObject(ctx).GetName()
```
I'm not sure how I feel about this. It gives every piece of the code access to the resource that's being reconciled, which is kind of nice. It may be a little bit of an abuse of the context mechanism. Happy to discuss.

Reply:
I agree injecting the entire object seems to be abusing the mechanism. Injecting the type.NamespacedName will give us a good starting point. As you suggested, now we have a unified injection package that handles options, restconfig and the resource namespaced name. It looks much cleaner this way.

Reply:
This injection package will also be the perfect place to implement a generic injection function that allows us to pass multiple labels down the call path (one for each node when there are multiple nodes in the defragmentation reconcile loop). It will be very useful later when we implement the Karpenter health metrics dashboard. In that case, there may be a need to let karpenter manage the injection key. We may pass the injection key through argument dynamically. Let's revisit this later when we come to the implementation details.